### DISCUSSION TITLE
Questions about input expressions

### DISCUSSION THREAD

**Discussion by yyy33** [houmain/keymapper#84] [Q&A] [answered]
Hi, I'm writing a mapping software for myself and then I'm referencing the keymapper idea.
I've been confused as to what the difference is between `()` and `{}`, I think that `{}` can do what `()` can do. For example, `A{B}` can be replaced by `(A)B` using

**Comment by houmain** [comment]
Did you see the tests in `test0_ParseKeySequence`? There you can see which ups and downs are generated from this syntax.
I doubt that it can be replaced. Do you mean on the input or the output side?

**Reply by yyy33** [reply]
I didn't look at `test0_ParseKeySequence`, I don't know c++, I'm just puzzled by the syntax from a user's point of view. I was referring to input expressions.

**Reply by houmain** [reply, marked as answer]
Do you really mean `(A)B`? Because that is the same as `A B` (`A` is in a group with only one key, which does not really make sense).
But this does not ensure that `A` is still hold when `B` is pressed, which `A{B}` does.

**Reply by yyy33** [reply]
I think I understand the difference between them.

**Reply by yyy33** [reply]
Do you think it would be easier to split each key in the input expression into press and release, I'm going to do that in my key mapping software

**Reply by houmain** [reply]
Easier?
For your implementation? Definitely.
For the user? Maybe for some use cases.
